megatron/data/gpt_dataset.py

```diff
@@ -114,8 +114,11 @@ def _build_indices():
                     print_rank_0('> elapsed time for building concat dataset indices: '
                                  '{:.2f} (sec)'.format(time.time() - start_time))
                     return dataset_index, dataset_sample_index
-                
+
                 self.dataset_index, self.dataset_sample_index = _build_indices()
+                np_rng = np.random.RandomState(seed=dataset_builders[0].seed)
+                self.shuffle_index = np.arange(self.num_samples)
+                np_rng.shuffle(self.shuffle_index)
```

Comment:
Create (and shuffle)[^shuffle] a global `shuffle_index` of `len(num_samples)`.

Explicitly, this

```python
BuildConcatDataset.shuffle_index
```

is a list of indices maps each sample to a particular {`dataset_index`, `dataset_sample_index`}

**Example**

```python
>>> import numpy as np
>>> shuffle_index = np.arange(10)
>>> np_rng = np.random.RandomState(seed=123)
>>> shuffle_index
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> np_rng.shuffle(shuffle_index)
>>> shuffle_index
array([4, 0, 7, 5, 8, 3, 1, 6, 9, 2])
```